KAFKA-17783: Adding listeners to remove share partition on partition changes

apoorvmittal10 opened the pull request:
The PR adds listeners to Partition class which might invoke the clean up of share partitions on Partition changes.

I considered introducing new Listener class which defines methods per SharePartitionKey but avoided as that would have required wiring up at the calls at different places in ReplicaManager. Additionally replica manager needed handling to store the new listener, etc. I found the current PR approach better as SharePartition is abstracted Shared view of Partition hence attaching the change events on Partition is helpful.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apoorvmittal10 committed:
KAFKA-17783: Adding listeners to remove share partition on partition changes

AndrewJSchofield reviewed (commented):
This looks like a promising approach. It's in an area of code that I really don't know, so I'm not qualified to a proper review. I do have a couple of comments:

* Can the partition become a leader after it has previously become a follower? I just wonder whether the state machine is more complicated than this PR implies?
* The listener is called under a lock in the `Partition`. The listener implementation needs to be very careful not to do anything brave under that lock. I expect there is potential for a deadlock here.

apoorvmittal10 commented:
Thanks @AndrewJSchofield for review and good points.

> Can the partition become a leader after it has previously become a follower? I just wonder whether the state machine is more complicated than this PR implies?

Yes, that can always happen. Once it does then next share fetch request should load the share partition. The scenario is not very different then partition being shuffled across brokers. The broker who is the leader of the partition should be able to load share partition.

> The listener is called under a lock in the Partition. The listener implementation needs to be very careful not to do anything brave under that lock. I expect there is potential for a deadlock here.

The lock is for each Partition, but I do get the point. I can make the call outside lock as well, I ll wait for @junrao's comments as he can help.

apoorvmittal10 committed:
Correcting comments

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-17783

mumrah commented:
@apoorvmittal10 what is the purpose of this listener? Is is so that SPM can clean up its in-memory state when it is no longer the leader for a partition?

If that's the main use case, I think we should consider tying into the metadata system directly rather than coupling ourselves to Partition. 

@dajac, since its looks like you added it, maybe you can comment on the intended usage of PartitionListener?

apoorvmittal10 commented:
> @apoorvmittal10 what is the purpose of this listener? Is is so that SPM can clean up its in-memory state when it is no longer the leader for a partition?

Yes, that's the purpose.


> If that's the main use case, I think we should consider tying into the metadata system directly rather than coupling ourselves to Partition.

So as far as I see the code there exists 1 approach taken by coordinators: in replica manager where we have a method `becomeLeaderOrFollower` where we pass `onLeadershipChange` method. The `onLeadershipChange` method triggers `groupCoordinator` and `txnCordinators`, `onElection` and `onResignation` methods. The 2) approach is by invoking the listeners on partition change.

I am not sure if we do get leadership change notifications from metadata system. Not sure if we could have some kind of leaderEpochChange listener to see the partition epoch bump and do some actions. But I found the listeners approach much sensible.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-17783

mumrah commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -665,17 +669,83 @@ public void handleFencedSharePartitionException(
+         */
+        @Override
+        public void onFollower(TopicPartition topicPartition) {
+            log.info("The share partition leader change listener is invoked for the topic-partition: {}, share-partition: {}",
```
INFO is probably too noisy for this as we expect it any time the topic leadership changes.

apoorvmittal10 replied:
Sure, can make it debug as well.

mumrah reviewed (commented):
Thanks for the details, @apoorvmittal10. As we discussed offline, those methods in ReplicaManager are driven by the metadata system (BrokerMetadataPublisher). In the end, if we're just acting on one partition at a time, it shouldn't matter if we use the partition listener or the metadata publisher. 

> I am not sure if we do get leadership change notifications from metadata system

Partition leadership, ISR, etc are all controller-managed metadata which flow through the metadata system. We can reliably learn if a partition became a leader or if it became a follower through MetadataPublisher (this is how ReplicaManager and the coordinators work).

--

Thinking a bit more about this, I'd like to use a metadata publisher rather than modifying the PartitionListener. It seems that PartitionListener is more geared towards the log state rather than the partition metadata. In fact, the only production usage of this listener I can find is kafka.coordinator.group.ListenerAdapter#onHighWatermarkUpdated. 

Since we are interested in subscribing to changes in the partition's leadership state, I think it is better to directly subscribe to the metadata system. This also has the benefit of eliminating the (potentially numerous) SharePartitionListener objects.

apoorvmittal10 commented:
Thanks @mumrah, works well with me. I was more thinking to tie it up with Partition listener initially but can work either way. @dajac @junrao What do you think?

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -82,6 +82,11 @@ trait PartitionListener {
+  /**
+   * Called when the Partition on this broker is marked as follower.
+   */
+  def onFollower(partition: TopicPartition): Unit = {}
```
If a broker switches from a follower to a follower, there is no need to remove the share partition. It's only when a broker is no longer a leader, a share partition should be removed. So we could add a more specific listener.

apoorvmittal10 replied:
Do you think either onLeaderToFollower or onBecomeFollower could be the better names?

apoorvmittal10 replied:
Done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the PR. Left a comment.

Regarding whether to use a listener at the Partition level or from BrokerMetadataPublisher, it's probably better to use a Partition level listener. It's true that the source of truth of the metadata is BrokerMetadataPublisher. However, the share partition requests only interact with metadata at the Partition level. So, it's more natural to take an action to share partitions after the metadata has been reflected at the Partition level. Otherwise, what could happen is that we remove a partition after the metadata changes in BrokerMetadataPublisher, but not in Partition yet. When we try to create the new SharePartition, it's still created with the outdated leaderEpoch.

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for the PR. Left a comment.
> 
> Regarding whether to use a listener at the Partition level or from BrokerMetadataPublisher, it's probably better to use a Partition level listener. It's true that the source of truth of the metadata is BrokerMetadataPublisher. However, the share partition requests only interact with metadata at the Partition level. So, it's more natural to take an action to share partitions after the metadata has been reflected at the Partition level. Otherwise, what could happen is that we remove a partition after the metadata changes in BrokerMetadataPublisher, but not in Partition yet. When we try to create the new SharePartition, it's still created with the outdated leaderEpoch.

Thanks a lot for confirming. Then I will go ahead with this approach and ll add tests to the PR.
cc: @mumrah @dajac

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-17783

apoorvmittal10 committed:
Correcting listener name, added tests

apoorvmittal10 commented:
@AndrewJSchofield @mumrah @junrao I have updated the code with tests. Also removed listener call from locked `markFollower` method, @AndrewJSchofield mentioned, rather moved to `replicaManager` itself. Please review.

chia7712 commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -609,6 +610,12 @@ private SharePartition getOrCreateSharePartition(SharePartitionKey sharePartitio
+                    // However, as there could be multiple share partitions (per group name) for a single topic-partition,
+                    // hence create separate listeners per share partition which holds the share partition key
+                    // to identify the respective share partition.
+                    replicaManager.maybeAddListener(sharePartitionKey.topicIdPartition().topicPartition(),
```
Do we have a mechanism to remove unused listeners? Since a listener is created for each group, the list of listeners in a partition could grow significantly over time. Especially considering that the listener includes a hot method `onHighWatermarkUpdated`

apoorvmittal10 replied:
Valid point, I have attached a removListener method on cleanup. Done.

chia7712 reviewed (commented):
@apoorvmittal10 thanks for this patch

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2455,6 +2455,8 @@ class ReplicaManager(val config: KafkaConfig,
             // Only change partition state when the leader is available
             if (partition.makeFollower(partitionState, highWatermarkCheckpoints, topicIds(partitionState.topicName))) {
+              // Invoke the follower transition listeners for the partition.
+              partition.invokeLeaderToFollowerListeners()
```
This is the ZK path, which won't be supported in 4.0.

apoorvmittal10 replied:
Thanks for that, I have removed it and validated in tests as well.

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -82,6 +82,11 @@ trait PartitionListener {
+  /**
+   * Called when the Partition on this broker is transitioned to follower from leader.
+   */
+  def onLeaderToFollower(partition: TopicPartition): Unit = {}
```
This method is invoke when a replica becomes a follower and may not be previously a leader. So, we could just name this method sth like `onBecomingFollower`.

apoorvmittal10 replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -6798,6 +6798,62 @@ class ReplicaManagerTest {
   }
 
+  @Test
+  def testBecomeFollowerInvokeLeaderToFollowerListener(): Unit = {
```
LeaderAndIsrRequest is used in ZK-based controller, which won't be supported in 4.0. We want to test KRaft-based controller logic instead.

apoorvmittal10 replied:
Agree, removed and added corrected tests to verify.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Left a few more comments.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
@junrao @chia7712 Thanks for reviewing the changes, I have addressed the comments.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -665,17 +670,78 @@ public void handleFencedSharePartitionException(
 
+    private static void removeSharePartitionFromCache(SharePartitionKey sharePartitionKey,
+        Map<SharePartitionKey, SharePartition> map, ReplicaManager replicaManager) {
+        SharePartition sharePartition = map.remove(sharePartitionKey);
```
Nit: A line break before the ) would separate the argument declarations from the statements more clearly.

apoorvmittal10 replied:
Done.

AndrewJSchofield reviewed (commented):
This is really not an area of code that I'm very familiar with. Just one small legibility comment. lgtm.

junrao commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -665,17 +670,78 @@ public void handleFencedSharePartitionException(
+
+        @Override
+        public void onBecomingFollower(TopicPartition topicPartition) {
+            log.debug("The share partition leader change listener is invoked for the topic-partition: {}, share-partition: {}",
```
leader change => becoming follower ?

apoorvmittal10 replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -6798,6 +6798,105 @@ class ReplicaManagerTest {
+  }
+
+  @Test
+  def testBecomeFollowerNotInvokeOnBecomingFollowerListenerOnZk(): Unit = {
```
No need to add this test since `replicaManager.becomeLeaderOrFollower` will be eventually removed.

apoorvmittal10 replied:
Done, removed.

junrao commented on core/src/test/java/kafka/server/share/SharePartitionManagerTest.java:
```diff
@@ -2496,6 +2498,98 @@ public void testReplicaManagerFetchMultipleSharePartitionsException() {
+        verify(sp0, times(1)).markFenced();
+        verify(mockReplicaManager, times(1)).removeListener(any(), any());
+
+        // Invoke listener for second share partition.
```
second share partition => non-matching share partition?

apoorvmittal10 replied:
Done.

junrao commented on core/src/test/java/kafka/server/share/SharePartitionManagerTest.java:
```diff
@@ -2496,6 +2498,98 @@ public void testReplicaManagerFetchMultipleSharePartitionsException() {
+
+        // Invoke listener for second share partition.
+        listenerConsumer.accept(tp);
+        // The second share partition should not be removed as the listener is attached to single topic partition.
```
attached to single topic partition => attached to a different topic partition ?

apoorvmittal10 replied:
Done.

junrao commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -665,17 +670,78 @@ public void handleFencedSharePartitionException(
+     * group and topic-partition. Instead of maintaining a separate map for topic-partition to share partitions,
+     * we can maintain the share partition key in the listener and create a new listener for each share partition.
+     */
+    static class SharePartitionListener implements PartitionListener {
```
Would it be better to make this a private class? This way, we don't need to pass in the states in SharePartitionManager.

apoorvmittal10 replied:
That's what I was initially thinking to do but it made writing tests hard as now I have to create an instance of SharePartitionManager and invoke calls to create SharePartition as well which will provide me the listener. Hence thought to make it static itself.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Just a few more minor comments.

chia7712 commented:
@apoorvmittal10 The listeners utilize a `CopyOnWriteArrayList`, meaning that both "add" and "remove" operations can trigger array copies. I am concerned that this could lead to performance issues when a partition becomes a follower, as it may result in `n * (n + 1) / 2` copies when listeners are removed one by one.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
> @apoorvmittal10 The listeners utilize a `CopyOnWriteArrayList`, meaning that both "add" and "remove" operations can trigger array copies. I am concerned that this could lead to performance issues when a partition becomes a follower, as it may result in `n * (n + 1) / 2` copies when listeners are removed one by one.

@chia7712 Yes, you are right it's `CopyOnWriteArrayList`. I considered that initially and my thought process was that `listeners (CopyOnWriteArrayList)` is per Partition. Hence a partition listener can contain listeners at most to the number of share groups on a broker. Yes, if the number of groups in really huge then it might cause a bit delay. Do you think that is a concern?
If yes, then I can have a single SharePartitionListener which keeps a mapping of `topicPartitions to set of sharePartitionKeys`.

junrao reviewed (approved):
@apoorvmittal10 : Thanks for the updated PR. LGTM.

Regarding `CopyOnWriteArrayList`, the size of the list is the number of groups. So, I agree that it's not worth optimizing.

chia7712 commented:
> Yes, if the number of groups in really huge then it might cause a bit delay. Do you think that is a concern?

In fact, I don't have strong evidence about the performance issue. I may run a benchmark for the scenario, but before conducting the experiment, I agree with @junrao's comment that the size is limited by the groups.

By the way, this patch is LGTM :)
